src/protocol.rs

```diff
@@ -0,0 +1,388 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! Logic for parsing and generating Quilkin Control Message Protocol (QCMP) messages.
+
+use nom::bytes::complete;
+// Magic number to distinguish control packets from regular traffic.
+const MAGIC_NUMBER: &[u8] = b"QLKN";
```

Comment:
Just for curiosity's sake - did you generate this from the compiler toolchain, or did you write this by hand?

I was playing around with the yaml in https://ide.kaitai.io/# just for fun, so was curious.

Reply:
@markmandel I wrote it by hand because as far as I could tell the Rust runtime for the kaitai compiler isn't maintained much currently. I think some people are reviving the effort but it's not quite there yet where we can rely on it. Eventually we can replace it with a generated implementation once it is more stable and reliable.

Reply:
@XAMPPRocky  I did notice that about the Rust runtime when I went and looked, but wasn't sure if you had found some magic 👍🏻  thanks!